### PR TITLE
tools: monstore: add 'show-versions' command.

### PR DESCRIPTION
**Pull request by dllgzcl**
Use this tool, we can list the first/last committed version of maps.

Signed-off-by: Cilang Zhao <zhao.cilang@h3c.com>

### CODE REVIEW

**Comment by dllgzcl** [comment]
Would you take a look and merge it ?@jecluis

**Comment by dllgzcl** [comment]
I have revised the commit, would you take a look?@jecluis @tchaikov

File: src/tools/ceph_monstore_tool.cc [hunk 1]
```diff
@@ -676,7 +678,39 @@ int main(int argc, char **argv) {
...
                 << std::endl;
     }
-  } else if (cmd == "dump-paxos") {
+  } else if (cmd == "show-versions") {
+    string map_type; //map type:osdmap,monmap...
+    // visible options for this command
+    po::options_description op_desc("Allowed 'show-versions' options");
+    op_desc.add_options()
+      ("help,h", "produce this help message")
+      ("map-type",po::value<string>(&map_type),"map_type");
```
Comment (tchaikov) [inline src/tools/ceph_monstore_tool.cc]: add a space after the comma ",",

File: src/tools/ceph_monstore_tool.cc [hunk 2]
```diff
@@ -676,7 +678,39 @@ int main(int argc, char **argv) {
...
+
+    if (op_vm.count("help") || map_type.empty()) {
+      usage(argv[0], op_desc);
+      err = 0;
+      goto done;
+    }
+
+    unsigned int v_first = 0;
+    unsigned int v_last = 0;
+    v_first = st.get(map_type,"first_committed");
```
Comment (tchaikov) [inline src/tools/ceph_monstore_tool.cc]: add a space after the comma.

File: src/tools/ceph_monstore_tool.cc [hunk 3]
```diff
@@ -676,7 +678,39 @@ int main(int argc, char **argv) {
...
+    v_last = st.get(map_type,"last_committed");
```
Comment (tchaikov) [inline src/tools/ceph_monstore_tool.cc]: ditto

File: src/tools/ceph_monstore_tool.cc [hunk 4]
```diff
@@ -676,7 +678,39 @@ int main(int argc, char **argv) {
...
+
+    std::cout << "first committed:\t" << v_first << "\r\n"
+              << "last  committed:\t" << v_last << std::endl;
+    } else if (cmd == "dump-paxos") {
```
Comment (tchaikov) [inline src/tools/ceph_monstore_tool.cc]: wrong indent.

File: src/tools/ceph_monstore_tool.cc [hunk 5]
```diff
@@ -676,7 +678,39 @@ int main(int argc, char **argv) {
...
+      err = 0;
+      goto done;
+    }
+
+    unsigned int v_first = 0;
+    unsigned int v_last = 0;
+    v_first = st.get(map_type,"first_committed");
+    v_last = st.get(map_type,"last_committed");
+
+    std::cout << "first committed:\t" << v_first << "\r\n"
```
Comment (tchaikov) [inline src/tools/ceph_monstore_tool.cc]: just out of curiosity, why would you put a "\r" here?

**Comment by dllgzcl** [comment]
I have revised the commit! Please take a look and merget it! @jecluis @tchaikov

**Comment by dllgzcl** [comment]
Would you merge it ? :) @tchaikov  @jecluis

**Comment by jecluis** [comment]
I've tested this and it works in its current form.

I noticed two things I think you should consider addressing on a later pull request though:

1. the tool will return success even if the map type does not exist (i.e., if <map_type>:first_committed does not exist, for instance); and

2. the output format is less than parseable. Granted it's not that difficult, but I'm sure a better format could be found. For instance, just writing out 'first_committed' or 'last_committed', instead of adding the extra space in-between words.

Anyway, I'll take it. :) Thanks!

**Pull request merged**